File: xiaozhi-esp32-server-main/main/manager-api/src/main/java/xiaozhi/modules/childprofile/support/InterestKeyNormalizer.java
package xiaozhi.modules.childprofile.support;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

import org.apache.commons.lang3.StringUtils;

public final class InterestKeyNormalizer {

    private static final Map<String, String> INTEREST_ALIASES = buildAliases();

    private InterestKeyNormalizer() {
    }

    public static String normalizeInterest(String value) {
        String raw = StringUtils.trimToEmpty(value);
        if (StringUtils.isBlank(raw)) {
            return "";
        }

        String normalized = normalizeAliasKey(raw);
        String mapped = INTEREST_ALIASES.get(normalized);
        return mapped != null ? mapped : raw;
    }

    public static List<String> normalizeInterestList(List<String> values) {
        if (values == null || values.isEmpty()) {
            return new ArrayList<>();
        }
        Set<String> deduped = new LinkedHashSet<>();
        for (String value : values) {
            String normalized = normalizeInterest(value);
            if (StringUtils.isNotBlank(normalized)) {
                deduped.add(normalized);
            }
        }
        return new ArrayList<>(deduped);
    }

    private static Map<String, String> buildAliases() {
        Map<String, String> aliases = new LinkedHashMap<>();

        register(aliases, "animals",
                "animals", "animal", "🐶小动物", "小动物", "动物", "动物们", "小动物们");
        register(aliases, "dinosaurs",
                "dinosaurs", "dinosaur", "🦖恐龙", "恐龙");
        register(aliases, "space",
                "space", "🚀太空", "太空", "宇宙", "星空");
        register(aliases, "vehicles",
                "vehicles", "vehicle", "🚗汽车和交通工具", "汽车和交通工具", "交通工具", "汽车", "车", "车车");
        register(aliases, "nature",
                "nature", "🌳大自然", "大自然", "自然", "户外自然");
        register(aliases, "sports",
                "sports", "sport", "⚽运动", "运动", "体育");
        register(aliases, "art_and_crafts",
                "art_and_crafts", "artandcrafts", "🎨画画和手工", "画画和手工", "画画", "手工", "美术");
        register(aliases, "music_and_dance",
                "music_and_dance", "musicanddance", "🎵音乐和跳舞", "音乐和跳舞", "音乐", "跳舞", "唱歌跳舞");
        register(aliases, "stories_and_picture_books",
                "stories_and_picture_books", "storiesandpicturebooks", "stories", "📚故事和绘本", "故事和绘本", "故事", "绘本", "图画书");
        register(aliases, "riddles_and_games",
                "riddles_and_games", "riddlesandgames", "games", "🧩猜谜和小游戏", "猜谜和小游戏", "猜谜", "小游戏", "游戏", "谜语");

        return aliases;
    }

    private static void register(Map<String, String> aliases, String key, String... values) {
        for (String value : values) {
            String normalized = normalizeAliasKey(value);
            if (StringUtils.isNotBlank(normalized)) {
                aliases.put(normalized, key);
            }
        }
    }

    private static String normalizeAliasKey(String value) {
        String normalized = StringUtils.trimToEmpty(value).toLowerCase(Locale.ROOT);
        normalized = normalized.replace("\uFE0F", "");
        normalized = normalized.replaceAll("[\\s_\\-]+", "");
        return normalized;
    }
}

File: main/xiaozhi-server/core/interest_key_normalizer.py
from __future__ import annotations

from typing import Iterable


INTEREST_ALIASES = {
    "animals": "animals",
    "animal": "animals",
    "小动物": "animals",
    "动物": "animals",
    "小动物们": "animals",
    "动物们": "animals",
    "🐶小动物": "animals",
    "dinosaurs": "dinosaurs",
    "dinosaur": "dinosaurs",
    "恐龙": "dinosaurs",
    "🦖恐龙": "dinosaurs",
    "space": "space",
    "太空": "space",
    "宇宙": "space",
    "星空": "space",
    "🚀太空": "space",
    "vehicles": "vehicles",
    "vehicle": "vehicles",
    "交通工具": "vehicles",
    "汽车和交通工具": "vehicles",
    "汽车": "vehicles",
    "车": "vehicles",
    "车车": "vehicles",
    "🚗汽车和交通工具": "vehicles",
    "nature": "nature",
    "大自然": "nature",
    "自然": "nature",
    "户外自然": "nature",
    "🌳大自然": "nature",
    "sports": "sports",
    "sport": "sports",
    "运动": "sports",
    "体育": "sports",
    "⚽运动": "sports",
    "artandcrafts": "art_and_crafts",
    "art_and_crafts": "art_and_crafts",
    "画画和手工": "art_and_crafts",
    "画画": "art_and_crafts",
    "手工": "art_and_crafts",
    "美术": "art_and_crafts",
    "🎨画画和手工": "art_and_crafts",
    "musicanddance": "music_and_dance",
    "music_and_dance": "music_and_dance",
    "音乐和跳舞": "music_and_dance",
    "音乐": "music_and_dance",
    "跳舞": "music_and_dance",
    "唱歌跳舞": "music_and_dance",
    "🎵音乐和跳舞": "music_and_dance",
    "storiesandpicturebooks": "stories_and_picture_books",
    "stories_and_picture_books": "stories_and_picture_books",
    "stories": "stories_and_picture_books",
    "故事和绘本": "stories_and_picture_books",
    "故事": "stories_and_picture_books",
    "绘本": "stories_and_picture_books",
    "图画书": "stories_and_picture_books",
    "📚故事和绘本": "stories_and_picture_books",
    "riddlesandgames": "riddles_and_games",
    "riddles_and_games": "riddles_and_games",
    "games": "riddles_and_games",
    "猜谜和小游戏": "riddles_and_games",
    "猜谜": "riddles_and_games",
    "小游戏": "riddles_and_games",
    "游戏": "riddles_and_games",
    "谜语": "riddles_and_games",
    "🧩猜谜和小游戏": "riddles_and_games",
}


def _normalize_alias_key(value: str | None) -> str:
    normalized = str(value or "").strip().lower()
    normalized = normalized.replace("\ufe0f", "")
    for char in (" ", "_", "-"):
        normalized = normalized.replace(char, "")
    return normalized


def normalize_interest_key(value: str | None) -> str:
    raw = str(value or "").strip()
    if not raw:
        return ""
    return INTEREST_ALIASES.get(_normalize_alias_key(raw), raw)


def normalize_interest_keys(values: Iterable[str] | None) -> list[str]:
    if not values:
        return []
    normalized: list[str] = []
    for value in values:
        key = normalize_interest_key(value)
        if key and key not in normalized:
            normalized.append(key)
    return normalized
